Use Authentication for ZeebeClient

stephanpelikan opened the pull request:
As mentioned in #558 we want to use spring-zeebe to connect to a self-managed cluster having tenants enabled. This also requires to enable Identity based authentication (see https://github.com/camunda/camunda-platform#multi-tenancy).

I used the current main-branch (https://github.com/camunda-community-hub/spring-zeebe/commit/17e1fc4d9109a46c8952f380755fa18b459e396b) to test this, but failed due to two lacks in implementation:

1. At https://github.com/camunda-community-hub/spring-zeebe/blob/17e1fc4d9109a46c8952f380755fa18b459e396b/spring-boot-starter-camunda/src/main/java/io/camunda/zeebe/spring/client/configuration/CommonClientConfiguration.java#L41 it is tested for `getCloud() != null` but there is always an object given.
2. At https://github.com/camunda-community-hub/spring-zeebe/blob/17e1fc4d9109a46c8952f380755fa18b459e396b/spring-boot-starter-camunda/src/main/java/io/camunda/zeebe/spring/client/properties/ZeebeClientConfigurationProperties.java#L766 the new authentication classes (like https://github.com/camunda-community-hub/spring-zeebe/blob/17e1fc4d9109a46c8952f380755fa18b459e396b/camunda-sdk-java/java-common/src/main/java/io/camunda/common/auth/SelfManagedAuthentication.java#L23) are not used for Zeebe-client authentication.

In this pull-request I resolved this by
1. Fixing CommonClientConfiguration
2. Detaching io.camunda.zeebe.client.ZeebeClientConfiguration and ZeebeClientConfigurationProperties
3. Using any other authentication than DefaultNoopAuthentication as CredentialsProvider for the ZeebeClient

With this in place I'm able to start self-managed C8 and use it with tenants enabled:

1. local C8 started like this: https://github.com/camunda/camunda-platform#multi-tenancy
2. Configured Zeebe-API to allow default-tenant
3. Use Spring config:
application.yaml:
```yaml
logging:
  level:
    io.camunda.zeebe.client.impl.ZeebeCallCredentials: ERROR

zeebe:
  client:
    broker:
      gateway-address: 127.0.0.1:26500
    security:
       plaintext: true

common:
  keycloak:
    url: http://localhost:18080
    realm: camunda-platform
  client-id: zeebe
  client-secret: zecret
```

1nb0und reviewed (commented):
Thank you for the PR! A couple of thoughts

- I like where you separated the ZeebeConfig from the ZeebeProperties. It makes it consistent with the rest of the clients as well. 
- The way you defined the application yaml is concise and is inlined to the goal of single Authentication across multiple modules thru the `common.` prefix

I'm good to merge this PR and I get the intent. But I have the following concerns:

  - We would need to preserve compatibility with the old way of authenticating with zeebe (i.e. without the `common.`). Not sure if its preserved but can you still do Self Managed Authentication without utilizing the `common.` properties?

```
zeebe:
  client:
    client-id: zeebe
    client-secret: zecret
    broker:
      gateway-address: 127.0.0.1:26500
    security:
       plaintext: true
```

Overall we would need to sort out the Authentication piece which is moving towards the upcoming Identity SDK. This could merge (with some clarifications), but knowing that its coming soon, most of the authentication refactoring here would either be no longer applicable. 

A recap:

> 1. Fixing CommonClientConfiguration

looks good!

> 2. Detaching io.camunda.zeebe.client.ZeebeClientConfiguration and ZeebeClientConfigurationProperties

this looks good too!

> 3. Using any other authentication than DefaultNoopAuthentication as CredentialsProvider for the ZeebeClient

need to look at this closely given this will change with upcoming Identity SDK

Let me know your thoughts.

stephanpelikan commented on spring-boot-starter-camunda/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientConfiguration.java:
```diff
@@ -0,0 +1,198 @@
+
+  @Override
+  public CredentialsProvider getCredentialsProvider() {
+    if (!(authentication instanceof DefaultNoopAuthentication)) {
+      return new CredentialsProvider() {
+        @Override
+        public void applyCredentials(Metadata headers) throws IOException {
+          final var authHeader = authentication.getTokenHeader(Product.ZEEBE);
+          final var authHeaderKey = Metadata.Key.of(authHeader.getKey(), Metadata.ASCII_STRING_MARSHALLER);
+          headers.put(authHeaderKey, authHeader.getValue());
+        }
+
+        @Override
+        public boolean shouldRetryRequest(Throwable throwable) {
+          return ((StatusRuntimeException) throwable).getStatus() == Status.DEADLINE_EXCEEDED;
+        }
+      };
+    }
```
We need to set a custom auth-url (KeyCloak) for the Zeebe connection which seems not to be implemented yet in 8.3.x what is to reason for introducing this wrapper. I simply tried to reuse what is prepared in `SelfManageAuthentication` since there is the `Product.ZEEBE` enum. I thought this is already prepared to be used in the near future, so I took this right now to get our scenario working. I didn't recognize what's the target picture for the authentication stuff added recently.

stephanpelikan commented on spring-boot-starter-camunda/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientConfiguration.java:
```diff
@@ -0,0 +1,198 @@
+        }
+      };
+    }
+    if (hasText(properties.getCloud().getClientId()) && hasText(properties.getCloud().getClientSecret())) {
+//        log.debug("Client ID and secret are configured. Creating OAuthCredientialsProvider with: {}", this);
+      return CredentialsProvider.newCredentialsProviderBuilder()
+        .clientId(properties.getCloud().getClientId())
+        .clientSecret(properties.getCloud().getClientSecret())
+        .audience(properties.getCloud().getAudience())
+        .scope(properties.getCloud().getScope())
+        .authorizationServerUrl(properties.getCloud().getAuthUrl())
+        .credentialsCachePath(properties.getCloud().getCredentialsCachePath())
+        .build();
+    }
+    if (Environment.system().get("ZEEBE_CLIENT_ID") != null && Environment.system().get("ZEEBE_CLIENT_SECRET") != null) {
+      // Copied from ZeebeClientBuilderImpl
+      OAuthCredentialsProviderBuilder builder = CredentialsProvider.newCredentialsProviderBuilder();
+      int separatorIndex = properties.getBroker().getGatewayAddress().lastIndexOf(58); //":"
+      if (separatorIndex > 0) {
+        builder.audience(properties.getBroker().getGatewayAddress().substring(0, separatorIndex));
+      }
+      return builder.build();
+    }
```
You may refactor this as you like having the target of the new authentications introduced in your mind. I was not my goal to keep the old way of authentication. I just wanted to add the bits necessary to make it work in our situation. I would appreciate if you could accept this for now because we need to go ahead with our Camunda based project. Of course I'm willing to support you, but our primary goal is to deliver our business software in time.

1nb0und replied:
@stephanpelikan I'm happy to work as well with you and really value your contributions. Looking forward to great collaborations ahead!

1nb0und replied:
@stephanpelikan `8.3.4.2` and `8.4.0-alpha2-rc3` has been released and should be publicly available shortly

stephanpelikan commented:
I rebased my branch due to a conflict in `CommonClientConfiguration.java` caused by this PR https://github.com/camunda-community-hub/spring-zeebe/pull/563. I dropped my changes in favor to the changes of the PR request since they also include the fix I did and I guess it is likely to see more changes until my PR is merged. The rest of the changes is still necessary.

1nb0und commented on spring-boot-starter-camunda/src/main/java/io/camunda/zeebe/spring/client/properties/ZeebeClientConfigurationProperties.java:
```diff
@@ -1,37 +1,27 @@
 
 @ConfigurationProperties(prefix = "zeebe.client")
-public class ZeebeClientConfigurationProperties implements ZeebeClientConfiguration {
+public class ZeebeClientConfigurationProperties {
```
I understand there is a need to decouple Configuration from Properties. But there is also a value in implementing this Properties class from `ZeebeClientConfiguration` and that is to ensure the client here implements in strict standards whats on zeebe-java-client

Are there other ways we can do the strictness checking now that we remove the inheritance as well?

1nb0und commented on spring-boot-starter-camunda/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientConfiguration.java:
```diff
@@ -0,0 +1,198 @@
+
+  @Override
+  public CredentialsProvider getCredentialsProvider() {
+    if (!(authentication instanceof DefaultNoopAuthentication)) {
```
I appreciate reverting the change. I'm happy to accept the PR, but would need to correct this condition so that users will not be stuck to always use override credentials headers in Zeebe. As it stands, there will be always Authentication object and it will not be `DefaultNoopAuthentication` provided they set the right properties.

My temporary plan / work-around would be to change the condition to check for flag `common.enabled` and then use this functionality. I'll make a release so we can have something usable for you but not breaking for others.

Expect this to be change substantially after that.

Thanks again!

1nb0und commented on spring-boot-starter-camunda/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientConfiguration.java:
```diff
@@ -0,0 +1,198 @@
+      return new CredentialsProvider() {
+        @Override
+        public void applyCredentials(Metadata headers) throws IOException {
+          final var authHeader = authentication.getTokenHeader(Product.ZEEBE);
+          final var authHeaderKey = Metadata.Key.of(authHeader.getKey(), Metadata.ASCII_STRING_MARSHALLER);
```
I'm curious, what is this `var` keyword? It sounds like a Kotlin syntax but it works on my end with plain Java.

1nb0und replied:
I see now, it sounds like it was introduced in [Java 10](https://www.geeksforgeeks.org/var-keyword-in-java/). Will probably refactor or replace this for consistency

stephanpelikan commented:
Thank you for finalizing (https://github.com/camunda-community-hub/spring-zeebe/pull/567), merging and releasing!
